### PR TITLE
Initial Client Functionality

### LINKED ISSUE #10: User can create a client
A user needs to be able to create a client, but what is that?  In this case it refers to an identity that will be attempting to get an id token for a user and potentially get more information about them and do things on their behalf.  In order to have a functioning request the client is needed.

## Basic fields for client
- **name** : a generic name to identify the client with users that the client can set
- **client_id** : a long, set of web-safe characters that uniquely identify the client (should be automatically generated by the server)
- **secret** : a long set of web-safe characters that serves as a challenge to authorize a client's request (should be automatically generated by the server)
- **redirect_uri** : an url endpoint which the client will specify auth code requests to be returned
- **authentication_method** : one of five different values http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication
  - client_secret_basic
  - client_secret_post
  - client_secret_jwt
  - none

### PR DESCRIPTION
**Pull request by benfalk**
This brings in the ablity for a user to create a client from
the web interface.  Currently, the fuctionality is limited; however
you can create a client and edit it later if you want.  If you set
the client type to "confidential" a secret is generated for the
client that can be used later.

The client secret has received some attention to secure the client
secret.  It is encrypted in a way which it can be retrieved to
display back to the user and it has a mirrored hash column that
will allow for fast lookups with the correct client id and secret.

**Note:** I have opted to leave out the option for clients to specify
what authentication method to use.  For now I plan to use the
default `client_secret_basic` since the others are optional.  This
will be revisited later.

closes #10

### CODE REVIEW

**Comment by benfalk** [comment]
## User's Client Page
my-client-page-blank

## Client Creation Form
client-creation-form

## Client Display
client-display-page

File: test/models/client_test.exs [hunk 1]
```diff
@@ -0,0 +1,98 @@
+defmodule Ganondorf.ClientTest do
+  use Ganondorf.ModelCase
+
+  @valid_attrs %{
+    name: "test app",
+    redirect_uris: ["https://example.com/cb"],
+    client_type: :confidential
+  }
+  @invalid_attrs %{}
```
Comment (edubkendo) [inline test/models/client_test.exs]: Would it be more clear to call these something like `required_attrs` `missing_required_attrs` ?

File: web/controllers/my/client_controller.ex [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  def action(conn, _) do
+    apply(__MODULE__, action_name(conn), [
+      conn,
+      conn.params,
+      Guardian.Plug.current_resource(conn)
+    ])
+  end
+
+  # TODO this was lifted strait from the authorization controller
```
Comment (edubkendo) [inline web/controllers/my/client_controller.ex]: Does this mean it's not working yet to prevent unauthorized users from accessing clients?

Comment (edubkendo) [inline web/controllers/my/client_controller.ex]: nvm, I see you just wrote tests that prove that isn't the case. Not sure I understand the comment then.

Comment (benfalk) [inline web/controllers/my/client_controller.ex]: Na, it means it needs to be refactored into a plug, it's in two places now.  Here is the issue I made for it, made more sense to do separately: https://github.com/three-triangles/ganondorf/issues/20

File: test/models/client_test.exs (hunk 1 above)
Comment (benfalk) [inline test/models/client_test.exs]: I like it, I'll change 'em.  Those are the default names the scaffolding add.

**Comment by edubkendo** [comment]
LGTM. Like the way `shouldi` gives you something like rspec's `contexts`.

**Pull request merged**